src/index.ts

```
@@ -1,32 +1,53 @@
-class User {
+interface IUser {
+  id?: number
   name: string
   favColor: string
   age: number
-    
-  constructor(name: string, color: string, age: number) {
-    this.name = name
-    this.favColor = color
-    this.age = age
-  }
 }
 
 class UserAPI {
-  users: Array<User>
+  private users: Array<IUser>
+  private nextId: number
 
   constructor() {
     this.users = []
+    this.nextId = 0
   } 
 
   private assignId() {
-
+    this.nextId++
+    return this.nextId
   }
 
-  addUser(user: User) {
-
+  addUser(user: IUser) {
+    const result = (this.users).find( ( { name, favColor, age }: IUser) => (name === user.name && favColor === user.favColor && age === user.age))
+    try {
+      if (user.id) {
+        throw "Error: User Id is set by database. Please resubmit without pre-existing value."
+      } else if (result) {
+        throw "Error: A user with those properties already exists in the database."
+      } 
+    } catch (error) {
+      return error
+    }
+
+    const { name, favColor, age } = user
+    const newUser = this.createUser(name, favColor, age)
+    this.users.push(newUser)
+    return newUser
   }
 
   getUserById(id: number) {
-
+    try {
+      const foundUser = this.users.find( user => user.id === id) 
+      if (foundUser) {
+        return foundUser
+      }
+      throw "Error: No user found."
+    }
+    catch (error) {
+      return (error)
```

we should not be returning errors ... let the function throw the error and have its caller deal with it

---

be sure to have a case in the console logs below that intentionalyl induces the error and catches for it there